server/market/market.go

```diff
@@ -587,6 +588,101 @@ func (m *Market) SubmitOrder(rec *orderRecord) error {
 	return <-m.SubmitOrderAsync(rec)
 }
 
+// processCancelOrderWhileSuspended is called when cancelling an order while
+// the market is suspended and Run is not runnning. The error sent on errChan
+// is returned to the client.
+//
+// This function:
+// 1. Removes the target order from the book.
+// 2. Unlocks the order coins.
+// 3. Updates the storage with the new cancel order and cancels the existing limit order.
+// 3. Responds to the client that the order was recieved.
+// 4. Sends the unbooked order to the order feeds.
+// 5. Creates a match object and sends it to the swapper to notify the client of the match.
+func (m *Market) processCancelOrderWhileSuspended(rec *orderRecord, errChan chan<- error) {
+	co, ok := rec.order.(*order.CancelOrder)
+	if !ok {
+		errChan <- ErrInvalidOrder
+		return
+	}
+
+	if cancelable, err := m.CancelableBy(co.TargetOrderID, co.AccountID); !cancelable {
+		errChan <- err
+		return
+	}
+
+	m.bookMtx.Lock()
+	lo, ok := m.book.Remove(co.TargetOrderID)
+	if !ok {
+		errChan <- ErrTargetNotCancelable
+		return
+	}
+	delete(m.settling, co.TargetOrderID)
+	m.bookMtx.Unlock()
+
+	m.unlockOrderCoins(lo)
+
+	sTime := time.Now().Truncate(time.Millisecond).UTC()
+	co.SetTime(sTime)
+
+	// Create the client response here, but don't send it until the order has been
+	// commited to the storage.
+	respMsg, err := m.orderResponse(rec)
+	if err != nil {
+		errChan <- fmt.Errorf("error creating order response: %v", err)
+		return
+	}
+
+	dur := int64(m.EpochDuration())
+	now := encode.UnixMilli(time.Now())
+	epochIdx := now / dur
+	if err := m.storage.NewArchivedCancel(co, epochIdx, int64(m.EpochDuration())); err != nil {
+		errChan <- err
+		return
+	}
+	if err := m.storage.CancelOrder(lo); err != nil {
+		errChan <- err
+		return
+	}
+
+	// If we fail to send the response, then we send this error at the end of
+	// the function. If we send the error here and return the function, then
+	// the match will not be recorded and the server DB will be in an inconsistent
+	// state.
+	sendResponseErr := m.auth.Send(rec.order.User(), respMsg)
+	if sendResponseErr != nil {
+		log.Errorf("error sending response: %v", err)
+	}
+
+	sig := &updateSignal{
+		action: unbookAction,
+		data: sigDataUnbookedOrder{
+			order:    lo,
+			epochIdx: 0,
+		},
+	}
+	m.sendToFeeds(sig)
+
+	match := order.Match{
+		Taker:    co,
+		Maker:    lo,
+		Quantity: lo.Remaining(),
+		Rate:     lo.Rate,
+		Epoch: order.EpochID{
+			Idx: uint64(epochIdx),
+			Dur: m.EpochDuration(),
+		},
+		FeeRateBase:  m.getFeeRate(m.Base(), m.baseFeeFetcher),
+		FeeRateQuote: m.getFeeRate(m.Quote(), m.quoteFeeFetcher),
+	}
+	if err := m.swapper.ProcessCancelMatchDuringSuspendedMarket(&match); err != nil {
```

Comment:
This is my only real issue with `processCancelOrderWhileSuspended`.  It's unfortunate that Swapper is presently handling all `match` requests even if they are not negotiating matches (see `toMonitor` in `(*Swapper).Negotiate`).  I get that the reason is that the handler for the response is a `Swapper` method (`processMatchAcks`), but there's no conceptual reason why the `Swapper` should be involved for this or any cancel orders AFAIK.

I don't want to recommend large refactoring for all cancels, but perhaps just for `processCancelOrderWhileSuspended` we can look into implementing a match ack handler in this package.  Note that the acks are basically worthless unless they signal the start of a swap negotiation, and not strictly necessary even in that case for a successful swap.  That wasn't always the case, and the reason it's still organized this way presently.  Although the match ack signature is stored, I don't think it's actually required anymore, even for swap matches.

Hence I think a barebones handler just needs to unmarshal a `msgjson.Acknowledgement` and log it for debugging, since there won't be any swap matches batched with the cancel match in this non-epoch match.

Reply:
I did think it was a bit strange that I was calling `Swapper` when there is no swap happening :).
I've added a new handler, let me know what you think.